src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java

```diff
@@ -330,51 +330,85 @@ public static void routeStaticNet(Net currNet,
-            visitedRoutingNodes.clear();
-            List<Node> pathNodes = new ArrayList<>();
+            int watchdog = 10000;
             Node node = sink.getConnectedNode();
-            if (debug) System.out.println(node);
-            LightweightRouteNode sinkRNode = RouterHelper.createRoutingNode(node, createdRoutingNodes);
-            sinkRNode.setPrev(null);
-            q.add(sinkRNode);
-            boolean success = false;
-            while (!q.isEmpty()) {
-                LightweightRouteNode routingNode = q.poll();
-                visitedRoutingNodes.add(routingNode);
-                if (debug) System.out.println("DEQUEUE:" + routingNode);
-                if (debug) System.out.println(", PREV = " + routingNode.getPrev() == null ? " null" : routingNode.getPrev());
-                if (success = isThisOurStaticSource(design, routingNode, netType, usedRoutingNodes)) {
-                    //trace back for a complete path
-                    if (debug) {
-                        System.out.println("SINK: TILE = " + sink.getTile().getName() + " NODE = " + sink.getConnectedNode().toString());
-                        System.out.println("SOURCE " + routingNode.toString() + " found");
+            if (usedRoutingNodes.contains(node)) {
+                sink.setRouted(true);
+            } else {
+                assert(prevNode.isEmpty());
+                prevNode.put(node, INVALID_NODE);
+                assert(q.isEmpty());
+                q.add(node);
+                search: while ((node = q.poll()) != null) {
+                    assert(!usedRoutingNodes.contains(node));
+                    assert(!(netType == NetType.VCC && node.isTiedToVcc()) || (netType == NetType.GND && node.isTiedToGnd()));
+
+                    if (isThisOurStaticSource(design, node, netType)) {
+                        break;
                     }
-                    while (routingNode != null) {
-                        usedRoutingNodes.add(routingNode);// use routed RNodes as the source
-                        pathNodes.add(routingNode.getNode());
 
-                        if (debug) System.out.println("  " + routingNode.toString());
-                        routingNode = routingNode.getPrev();
+                    for (Node uphillNode : node.getAllUphillNodes()) {
+                        if (routeThruHelper.isRouteThru(uphillNode, node)) {
+                            continue;
+                        }
+
+                        switch(uphillNode.getIntentCode()) {
+                            case NODE_GLOBAL_VDISTR:
+                            case NODE_GLOBAL_HROUTE:
+                            case NODE_GLOBAL_HDISTR:
+                            case NODE_HLONG:
+                            case NODE_VLONG:
+                            case NODE_GLOBAL_VROUTE:
+                            case NODE_GLOBAL_LEAF:
+                            case NODE_GLOBAL_BUFG:
+                                continue;
+                        }
```

Comment:
Interesting construct, is this faster than an EnumSet?

Reply:
I've no idea actually! This chunk code was inlined from a related method, and there doesn't seem to be a lot of hits on the internet looking at this exact question.